<p style="align: center;"><img src="https://static.tildacdn.com/tild6636-3531-4239-b465-376364646465/Deep_Learning_School.png", width=300, height=300></p>

<h3 style="text-align: center;"><b>Физтех-Школа Прикладной математики и информатики (ФПМИ) МФТИ</b></h3>

<h2 style="text-align: center;"><b>Библиотека <a href="http://pandas.pydata.org/">pandas</a></b></h2>

Библиотека `pandas` активно используется в современном data science для работы с данными, которые могут быть представлены в виде таблиц (а это очень, очень большая часть данных)

`pandas` есть в пакете Anaconda, но если вдруг у Вас её по каким-то причинам нет, то можно установить, раскомментировав следующую команду:

In [1]:
#!pip install pandas

In [2]:
import pandas as pd
import numpy as np

In [3]:
np_array = np.random.randint(low=-15, high=20, size=4)
print(np_array)

[-10  -5 -10   2]


In [4]:
pd_series_1 = pd.Series(np_array)
pd_series_1

0   -10
1    -5
2   -10
3     2
dtype: int64

In [7]:
pd_series_2 = pd.Series(np_array, index = ['1-st day', '2-nd day', '3-rd day', '4-th day'])
print(pd_series_2)

1-st day   -10
2-nd day    -5
3-rd day   -10
4-th day     2
dtype: int64


In [8]:
# Таким образом можно обращаться по индексу
pd_series_2['4-th day']

2

In [10]:
pd_series_3 = pd.Series(np_array, index=['day1','day2','day3','day4'], name = 'Temperature')
print(pd_series_3)

day1   -10
day2    -5
day3   -10
day4     2
Name: Temperature, dtype: int64


In [15]:
# с индексами можно работать, как с обычными листами, т.е. обращаться по значению, брать слайс, или делать обратный слайс
print(pd_series_3[0])
print('*'*10)
print(pd_series_3[1:3])
print('*'*10)
print(pd_series_3[::-1])

-10
**********
day2    -5
day3   -10
Name: Temperature, dtype: int64
**********
day4     2
day3   -10
day2    -5
day1   -10
Name: Temperature, dtype: int64


In [16]:
# сортировка по индексу
pd_series_3.sort_index()

day1   -10
day2    -5
day3   -10
day4     2
Name: Temperature, dtype: int64

In [17]:
# сортировка по значению
pd_series_3.sort_values()

day1   -10
day3   -10
day2    -5
day4     2
Name: Temperature, dtype: int64

In [20]:
# series очень похож на стандартный np массив, можно прибавлять что-то, можно среднее брать, стандартное отклонение и прочее
# можно применять функции из numpy
pd_series_3.mean()
pd_series_3 + 10
pd_series_3.std()

5.678908345800274

In [25]:
# не важно в каком порядке лежат данные в series // dataframe - складывается все по индексу
pd_series_1 = pd.Series(np.random.randint(low=-10, high=5, size=5), index=['1','2','3','4','5'])
pd_series_2 = pd.Series(np.random.randint(low=-5, high=10, size=5), index=['1','2','3','4','5'])
pd_series_3 = pd_series_2 + pd_series_1
print(pd_series_1, '\n', '*'*5)
print(pd_series_2, '\n', '*'*5)
print(pd_series_3, '\n', '*'*5)

1   -1
2    2
3   -8
4    0
5   -4
dtype: int64 
 *****
1    3
2   -1
3    4
4    0
5    3
dtype: int64 
 *****
1    2
2    1
3   -4
4    0
5   -1
dtype: int64 
 *****


In [29]:
# еще один способ задавать series
pd_series_1 = pd.Series(np.random.randint(1,10,5))
pd_series_2 = pd.Series(np.random.randint(1,5,4))
print(pd_series_1,'\n')
print(pd_series_2,'\n')
pd_series_3 = pd_series_2 + pd_series_1
pd_series_3.shape
print(pd_series_3,'\n')

0    8
1    8
2    4
3    8
4    2
dtype: int64 

0    1
1    2
2    1
3    3
dtype: int64 

0     9.0
1    10.0
2     5.0
3    11.0
4     NaN
dtype: float64 



In [31]:
date_range = pd.date_range('20200601', periods=10)
print(date_range)

DatetimeIndex(['2020-06-01', '2020-06-02', '2020-06-03', '2020-06-04',
               '2020-06-05', '2020-06-06', '2020-06-07', '2020-06-08',
               '2020-06-09', '2020-06-10'],
              dtype='datetime64[ns]', freq='D')


In [33]:
pd_series_4 = pd.Series(np.random.randint(1,10,10), index = date_range)
print(pd_series_4)

2020-06-01    9
2020-06-02    4
2020-06-03    8
2020-06-04    1
2020-06-05    1
2020-06-06    5
2020-06-07    8
2020-06-08    1
2020-06-09    3
2020-06-10    9
Freq: D, dtype: int64


In [34]:
# как и в numpy можно получать булевые массивы
pd_series_4 > 3

2020-06-01     True
2020-06-02     True
2020-06-03     True
2020-06-04    False
2020-06-05    False
2020-06-06     True
2020-06-07     True
2020-06-08    False
2020-06-09    False
2020-06-10     True
Freq: D, dtype: bool

In [36]:
# объединение диапазонов значений
pd_series_4[(pd_series_4>4) & (pd_series_4<8)]

2020-06-06    5
Freq: D, dtype: int64

In [51]:
# DataFrame
d =  {
        'One': pd.Series(np.random.randint(1,10,3), index = (['a','b','c'])),
        'Two': pd.Series(np.random.randint(1,20,4), index= (['a','b','c','d']))
     }
type(d)
df = pd.DataFrame(d)
print(df)

   One  Two
a  3.0    5
b  9.0   13
c  3.0    7
d  NaN    6


In [52]:
df.index

Index(['a', 'b', 'c', 'd'], dtype='object')

In [53]:
# Обращение к одной из колонок Dataframe дает Series, после этого можно по индексу Series  обратиться к значению
print(type(df['One']))
df['One']['c']

3.0

In [55]:
df.columns

Index(['One', 'Two'], dtype='object')

In [61]:
type(df['One'])

pandas.core.series.Series

In [62]:
# к колонкам можно обращаться через точку, если нет пробелов в названии колонки
df.One

a    3.0
b    9.0
c    3.0
d    NaN
Name: One, dtype: float64

In [71]:
# можно делать слайс по строчкам, питон такое обращение поймет и распарсит
df[1:3]

One  Two
b  9.0   13
c  3.0    7

In [73]:
df['Three'] = df['One']+df['Two']
print(df['Three'])

a     8.0
b    22.0
c    10.0
d     NaN
Name: Three, dtype: float64


In [82]:
df_1 = pd.DataFrame(df, index = ['a','b','c','d'])
df_2 = pd.DataFrame({'One': [1,2]}, index = ['e','f'])
df_3 = pd.concat([df_1, df_2], sort=False)
print(df_3)

   One   Two  Three
a  3.0   5.0    8.0
b  9.0  13.0   22.0
c  3.0   7.0   10.0
d  NaN   6.0    NaN
e  1.0   NaN    NaN
f  2.0   NaN    NaN


In [93]:
# shift + tab x2 показывает какие аргументы принимает функция
pd_series = pd.Series(np.random.randint(low = -10, high = 5, size = 5, dtype = 'int'), index=range(0,5))
pd_dataframe = pd.DataFrame(pd_series)
pd_dataframe.columns = ['col_1']
print(pd_dataframe)

   col_1
0      4
1     -8
2     -5
3     -3
4     -2


Нужны данные. Про AppStore слышали? :)

### Скачать данные: https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps

Нас интересует файл `AppStore.csv`.

Кстати, `.csv` (Comma Separated Values) - наверное, самый частый формат данных в современном data science. По сути, это то же самое, что и `.xls` (и `.xlsx`), то есть таблица. 

Итак, *чтение файла с данными*:

In [137]:
import pandas as pd
import numpy as np
data = pd.read_csv('./app-store-apple-data-set-10k-apps/AppleStore.csv')

Посмотрим, что же такое переменная `data`:

In [53]:
data.sample(5)

Unnamed: 0          id                    track_name  size_bytes  \
1895        2204   632079234                  Star Command   153100288   
4065        5052  1020327340                DragonSoul RPG   212710400   
3653        4453   982354972                   Lifeline...    29690880   
4998        6546  1068360791                       謎解き2016    15844352   
6878       10252  1154791600  アウト～まさか浮気してないよね？～【恋愛不倫探しゲーム】    40107008   

     currency  price  rating_count_tot  rating_count_ver  user_rating  \
1895      USD   2.99              3473               619          4.0   
4065      USD   0.00              7681                15          4.5   
3653      USD   1.99             21347              1187          4.5   
4998      USD   0.00                 0                 0          0.0   
6878      USD   0.00                 0                 0          0.0   

      user_rating_ver     ver cont_rating prime_genre  sup_devices.num  \
1895              4.0   1.1.4         12+       Games               43   
4065              3.0  2.13.0          9+       Games               38   
3653              4.5  1.10.1          9+       Games               37   
4998              0.0     1.3          4+        Book               38   
6878              0.0   1.0.5          9+        Book               37   

      ipadSc_urls.num  lang.num  vpp_lic  
1895                5         1        1  
4065                5        17        1  
3653                5         6        1  
4998                0         1        1  
6878                0         1        1

In [54]:
type(data)

pandas.core.frame.DataFrame

Тип == `pandas.core.frame.DataFrame`, обычно говорят просто "датафрейм", то есть "кусок данных".

Можно вывести информацию по датафрейму:

In [136]:
data = pd.read_csv('./app-store-apple-data-set-10k-apps/AppleStore.csv', index_col='id')
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7197 entries, 281656475 to 1188375727
Data columns (total 16 columns):
Unnamed: 0          7197 non-null int64
track_name          7197 non-null object
size_bytes          7197 non-null int64
currency            7197 non-null object
price               7197 non-null float64
rating_count_tot    7197 non-null int64
rating_count_ver    7197 non-null int64
user_rating         7197 non-null float64
user_rating_ver     7197 non-null float64
ver                 7197 non-null object
cont_rating         7197 non-null object
prime_genre         7197 non-null object
sup_devices.num     7197 non-null int64
ipadSc_urls.num     7197 non-null int64
lang.num            7197 non-null int64
vpp_lic             7197 non-null int64
dtypes: float64(3), int64(8), object(5)
memory usage: 955.9+ KB


И более статистически значимую информацию:

In [56]:
print(data.shape)
data.describe()

(7197, 16)


Unnamed: 0    size_bytes        price  rating_count_tot  \
count   7197.000000  7.197000e+03  7197.000000      7.197000e+03   
mean    4759.069612  1.991345e+08     1.726218      1.289291e+04   
std     3093.625213  3.592069e+08     5.833006      7.573941e+04   
min        1.000000  5.898240e+05     0.000000      0.000000e+00   
25%     2090.000000  4.692275e+07     0.000000      2.800000e+01   
50%     4380.000000  9.715302e+07     0.000000      3.000000e+02   
75%     7223.000000  1.819249e+08     1.990000      2.793000e+03   
max    11097.000000  4.025970e+09   299.990000      2.974676e+06   

       rating_count_ver  user_rating  user_rating_ver  sup_devices.num  \
count       7197.000000  7197.000000      7197.000000      7197.000000   
mean         460.373906     3.526956         3.253578        37.361817   
std         3920.455183     1.517948         1.809363         3.737715   
min            0.000000     0.000000         0.000000         9.000000   
25%            1.000000     3.500000         2.500000        37.000000   
50%           23.000000     4.000000         4.000000        37.000000   
75%          140.000000     4.500000         4.500000        38.000000   
max       177050.000000     5.000000         5.000000        47.000000   

       ipadSc_urls.num     lang.num      vpp_lic  
count      7197.000000  7197.000000  7197.000000  
mean          3.707100     5.434903     0.993053  
std           1.986005     7.919593     0.083066  
min           0.000000     0.000000     0.000000  
25%           3.000000     1.000000     1.000000  
50%           5.000000     1.000000     1.000000  
75%           5.000000     8.000000     1.000000  
max           5.000000    75.000000     1.000000

Итак, наши данные -- это какая-то **информация про приложения в AppStore**. Давайте посмотрим, что тут есть:

In [57]:
# все столбцы датафрейма (названия)
data.columns

Index(['Unnamed: 0', 'track_name', 'size_bytes', 'currency', 'price',
       'rating_count_tot', 'rating_count_ver', 'user_rating',
       'user_rating_ver', 'ver', 'cont_rating', 'prime_genre',
       'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'],
      dtype='object')

In [142]:
data[data['user_rating']==data['user_rating'].max()]
data.sort_values('user_rating', ascending=False)
data[(data['user_rating']>4) & (data['user_rating']<5)]

(data['user_rating'] == data.user_rating.max()).sum()

492

In [164]:
import numpy as np
len(set(data['prime_genre']))
len(np.unique(data['prime_genre']))
data.prime_genre.unique()
df_counts = data['prime_genre'].value_counts()
data['prime_genre'].value_counts()
# data['price'].value_counts()

Games                3862
Entertainment         535
Education             453
Photo & Video         349
Utilities             248
Health & Fitness      180
Productivity          178
Social Networking     167
Lifestyle             144
Music                 138
Shopping              122
Sports                114
Book                  112
Finance               104
Travel                 81
News                   75
Weather                72
Reference              64
Food & Drink           63
Business               57
Navigation             46
Medical                23
Catalogs               10
Name: prime_genre, dtype: int64

In [175]:
data['prime_genre'].value_counts().max()
print(data.columns)  
data[data['prime_genre']=='Finance']['size_bytes'].sum()

Index(['Unnamed: 0', 'id', 'track_name', 'size_bytes', 'currency', 'price',
       'rating_count_tot', 'rating_count_ver', 'user_rating',
       'user_rating_ver', 'ver', 'cont_rating', 'prime_genre',
       'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'],
      dtype='object')


8136529366

In [184]:
data[(data['price']==0) & (data['user_rating']>3)]['id'].count()

2944

In [185]:
data[data.price > 100]

Unnamed: 0         id                      track_name  size_bytes  \
115          129  308368164  Proloquo2Go - Symbol-based AAC   723764224   
1479        1714  551215116             LAMP Words For Life   583263232   

     currency   price  rating_count_tot  rating_count_ver  user_rating  \
115       USD  249.99               773                10          4.0   
1479      USD  299.99                41                 0          4.0   

      user_rating_ver    ver cont_rating prime_genre  sup_devices.num  \
115               3.5  5.0.1          4+   Education               37   
1479              0.0  1.5.5          4+   Education               24   

      ipadSc_urls.num  lang.num  vpp_lic  
115                 5         3        1  
1479                5         1        1

---

#### Задание 0

О чём данные? (судя по столбцам)

---

Все значения датафрейма:

In [58]:
# одна строка == описание одного объекта, в данном случае -- приложения из AppStore
# data.values - конвертирует DataFrame в numpy массив
data.values

array([[1, 'PAC-MAN Premium', 100788224, ..., 5, 10, 1],
       [2, 'Evernote - stay organized', 158578688, ..., 5, 23, 1],
       [3, 'WeatherBug - Local Weather, Radar, Maps, Alerts', 100524032,
        ..., 5, 3, 1],
       ...,
       [11087, 'Bret Michaels Emojis + Lyric Keyboard', 111322112, ...,
        1, 1, 1],
       [11089, 'VR Roller Coaster World - Virtual Reality', 97235968,
        ..., 0, 2, 1],
       [11097, 'Escape the Sweet Shop Series', 90898432, ..., 0, 2, 1]],
      dtype=object)

In [9]:
data.values[0]

array([1, 'PAC-MAN Premium', 100788224, 'USD', 3.99, 21292, 26, 4.0, 4.5,
       '6.3.5', '4+', 'Games', 38, 5, 10, 1], dtype=object)

In [81]:
# data.loc[0,['price']]
# data.loc[0:21,['price']]
# data.loc[lambda data: data['price']>=100]
data.loc[(data['price']>=100) & (data['price']<=250)]
data.loc[0:2,['price']]
data.iloc[:100,:5]
data['price'][281656475]

3.99

In [82]:
data.shape

(7197, 16)

-- то есть 7197 строк и 17 столбцов.

Важно уметь обращаться к определённой строке или столбцу dataframe'а. Индексирование очень похоже на numpy, однако есть свои тонкости:

* Получить весь столбец (в даном случае столбец 'track_name'):

In [83]:
data['track_name']

id
281656475                                       PAC-MAN Premium
281796108                             Evernote - stay organized
281940292       WeatherBug - Local Weather, Radar, Maps, Alerts
282614216     eBay: Best App to Buy, Sell, Save! Online Shop...
282935706                                                 Bible
                                    ...                        
1187617475                                                Kubik
1187682390                                    VR Roller-Coaster
1187779532                Bret Michaels Emojis + Lyric Keyboard
1187838770            VR Roller Coaster World - Virtual Reality
1188375727                         Escape the Sweet Shop Series
Name: track_name, Length: 7197, dtype: object

In [84]:
type(data['track_name'])

pandas.core.series.Series

In [85]:
data['track_name'][:3]

id
281656475                                    PAC-MAN Premium
281796108                          Evernote - stay organized
281940292    WeatherBug - Local Weather, Radar, Maps, Alerts
Name: track_name, dtype: object

`pandas.core.series.Series` -- это тип подвыборки датафрейма. С этим типом работают реже, но стоит знать, что он есть.

* Получить определённую строку:

In [86]:
# iloc позволяет обращаться в DataFrame по номерам строк и колонок
data.iloc[0,10]

'4+'

In [118]:
data = pd.read_csv('./app-store-apple-data-set-10k-apps/AppleStore.csv', index_col=0)

In [119]:
data.describe()
# data.iloc[1,:]
# data.loc[1,['track_name', 'price']]
# data.index
# data.iloc[1,1]

id    size_bytes        price  rating_count_tot  \
count  7.197000e+03  7.197000e+03  7197.000000      7.197000e+03   
mean   8.631310e+08  1.991345e+08     1.726218      1.289291e+04   
std    2.712368e+08  3.592069e+08     5.833006      7.573941e+04   
min    2.816565e+08  5.898240e+05     0.000000      0.000000e+00   
25%    6.000937e+08  4.692275e+07     0.000000      2.800000e+01   
50%    9.781482e+08  9.715302e+07     0.000000      3.000000e+02   
75%    1.082310e+09  1.819249e+08     1.990000      2.793000e+03   
max    1.188376e+09  4.025970e+09   299.990000      2.974676e+06   

       rating_count_ver  user_rating  user_rating_ver  sup_devices.num  \
count       7197.000000  7197.000000      7197.000000      7197.000000   
mean         460.373906     3.526956         3.253578        37.361817   
std         3920.455183     1.517948         1.809363         3.737715   
min            0.000000     0.000000         0.000000         9.000000   
25%            1.000000     3.500000         2.500000        37.000000   
50%           23.000000     4.000000         4.000000        37.000000   
75%          140.000000     4.500000         4.500000        38.000000   
max       177050.000000     5.000000         5.000000        47.000000   

       ipadSc_urls.num     lang.num      vpp_lic  
count      7197.000000  7197.000000  7197.000000  
mean          3.707100     5.434903     0.993053  
std           1.986005     7.919593     0.083066  
min           0.000000     0.000000     0.000000  
25%           3.000000     1.000000     1.000000  
50%           5.000000     1.000000     1.000000  
75%           5.000000     8.000000     1.000000  
max           5.000000    75.000000     1.000000

In [120]:
# Как много по времени займет подготовка статьи?
# На каких конференциях лучше публиковаться по NLP и маршрутизации?
# Когда лучше начинать подготовку, как думаешь?
# По данным сравнение - лучше на открытых?
# Как ты думаешь, реально будет выступить на конференции в этом году, если статью примут или из-за ковид все сдвинется?

print(data.columns)
data.loc[:,['track_name','price','user_rating']].sort_index()
# data['best'] = map(lambda x,y: y/x if x != 0 else y, data['user_rating'], data['price'])
# data['best'] = data['price']/data['user_rating'] if data['user_rating'] != 0 else data['price']

Index(['id', 'track_name', 'size_bytes', 'currency', 'price',
       'rating_count_tot', 'rating_count_ver', 'user_rating',
       'user_rating_ver', 'ver', 'cont_rating', 'prime_genre',
       'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'],
      dtype='object')


track_name  price  user_rating
1                                        PAC-MAN Premium   3.99          4.0
2                              Evernote - stay organized   0.00          4.0
3        WeatherBug - Local Weather, Radar, Maps, Alerts   0.00          3.5
4      eBay: Best App to Buy, Sell, Save! Online Shop...   0.00          4.0
5                                                  Bible   0.00          4.5
...                                                  ...    ...          ...
11081                                              Kubik   0.00          4.5
11082                                  VR Roller-Coaster   0.00          4.5
11087              Bret Michaels Emojis + Lyric Keyboard   1.99          4.5
11089          VR Roller Coaster World - Virtual Reality   0.00          4.5
11097                       Escape the Sweet Shop Series   0.00          5.0

[7197 rows x 3 columns]

Более продвинутое и полезное индексирование:

In [15]:
data.iloc[5, 10]

'4+'

-- `data.iloc[i, j]`, где `i` -- номер строки, `j` -- номер столбца

Почти все типы индексирования:

* к строкам по числовому индексу, к столбцам -- по имени

In [32]:
data.loc[[0, 1, 2], ['track_name', 'id']]

track_name         id
0                                  PAC-MAN Premium  281656475
1                        Evernote - stay organized  281796108
2  WeatherBug - Local Weather, Radar, Maps, Alerts  281940292

* к строкам и к столбцам по числовому индексу

In [34]:
data.iloc[[1, 3, 5], [2, 1]]

track_name         id
1                          Evernote - stay organized  281796108
3  eBay: Best App to Buy, Sell, Save! Online Shop...  282614216
5                                   Shanghai Mahjong  283619399

* поддерживает и то, и то, но **запрещён**:

In [36]:
data.ix[[2,1,0], ['track_name', 'id']]

c:\users\izakharkin\anaconda3\envs\vision\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


track_name         id
2  WeatherBug - Local Weather, Radar, Maps, Alerts  281940292
1                        Evernote - stay organized  281796108
0                                  PAC-MAN Premium  281656475

In [38]:
data.ix[[0,1,2], [2, 1]]

c:\users\izakharkin\anaconda3\envs\vision\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


track_name         id
0                                  PAC-MAN Premium  281656475
1                        Evernote - stay organized  281796108
2  WeatherBug - Local Weather, Radar, Maps, Alerts  281940292

Точно также дело обстоит со **срезами**:

In [42]:
data.iloc[5:100, 0:5]

Unnamed: 0         id                                         track_name  \
5            6  283619399                                   Shanghai Mahjong   
6            7  283646709             PayPal - Send and request money safely   
7            8  284035177                            Pandora - Music & Radio   
8            9  284666222                        PCalc - The Best Calculator   
9           10  284736660                                        Ms. PAC-MAN   
10          11  284791396                          Solitaire by MobilityWare   
11          12  284815117                                   SCRABBLE Premium   
12          13  284815942               Google – Search made just for mobile   
13          14  284847138                   Bank of America - Mobile Banking   
14          15  284862767                                           FreeCell   
15          16  284876795             TripAdvisor Hotels Flights Restaurants   
16          17  284882215                                           Facebook   
17          18  284910350     Yelp - Nearby Restaurants, Shopping & Services   
18          20  284993459  Shazam - Discover music, artists, videos & lyrics   
19          21  285005463                      Crash Bandicoot Nitro Kart 3D   
20          22  285946052                                             iQuran   
21          23  285994151                                        :) Sudoku +   
22          24  286058814    Yahoo Sports - Teams, Scores, News & Highlights   
23          25  286070473                          Mileage Log | Fahrtenbuch   
24          27  286799607                        Cleartune - Chromatic Tuner   
25          28  286906691          Lifesum – Inspiring healthy lifestyle app   
26          29  286911400                                           Hangman.   
27          31  288113403      iTranslate - Language Translator & Dictionary   
28          32  288120394                                           TouchOSC   
29          33  288419283                                         RadarScope   
30          34  288429040                                           LinkedIn   
31          35  289084315                              Period Tracker Deluxe   
32          36  289446241                                  Election 2016 Map   
33          37  289523017                          Blackjack by MobilityWare   
34          39  289894882                                        White Noise   
..         ...        ...                                                ...   
70          80  299853944                                新浪新闻-阅读最新时事热门头条资讯视频   
71          81  299949744                                        MotionX GPS   
72          82  300048137                     AccuWeather - Weather for Life   
73          83  300238550      Mint: Personal Finance, Budget, Bills & Money   
74          84  300255638            ABC News - US & World News + Live Video   
75          85  300265786                    Rowmote: Remote Control for Mac   
76          86  300590611                                      Peekaboo Barn   
77          87  300704847                                 Speedtest by Ookla   
78          89  301387274                                         Pocket God   
79          90  301521403     Nike+ Training Club - Workouts & Fitness Plans   
80          91  301724680                                       Citi Mobile®   
81          92  302049354                  Phase 10 Pro - Play Your Friends!   
82          93  302584613        Kindle – Read eBooks, Magazines & Textbooks   
83          95  303058767                                               MASH   
84          96  303191318                                          同花顺-炒股、股票   
85          97  303337299                                Wedding Dash Deluxe   
86          99  303849934                                     Beer Pong Game   
87         100  304731501                    Tempo - Metronome with Setlists   
88  

Но обычно нужно уметь отвечать на более осмысленные вопросы, например:

* сколько приложений имеют рейтинг юзеров >= 4?

In [53]:
len(data[data['user_rating'] >= 4])

4781

* какие есть валюты в столбце currency?

In [52]:
np.unique(data['currency'])

array(['USD'], dtype=object)

* сколько приложений жанра Games? А тех, которые Games и рейтинг >= 4?

In [55]:
# 1
len(data[data['prime_genre'] == 'Games'])

3862

In [56]:
# 2
len(data[(data['user_rating'] >= 4) & (data['prime_genre'] == 'Games')])

2868

Оператор "ИЛИ" аналогично &, только он через |

In [57]:
len(data[(data['user_rating'] >= 4) | (data['prime_genre'] == 'Games')])

5775

---

#### Задание 1

Получите ответ на следующие вопросы (с помощью pandas):

1). Сколько всего жанров и какие они?  
2). Сколько суммарно байт весят все приложения жанра Finance?  
3). Сколько всего бесплатных приложений, у которых рейтинг юзеров больше трёх?  
4). Есть ли приложения, которые стоят больше 100 долларов?

In [ ]:
# Ваш код здесь

---

Часто надо сделать что-то такое:

In [19]:
import numpy as np

In [20]:
np.array(data['size_bytes'])

array([100788224, 158578688, 100524032, ..., 111322112,  97235968,
        90898432], dtype=int64)

То есть взять столбец из таблицы и засунуть его в np.array(). Это можно сделать и просто обратившись к .values, ведь его тип и есть np.array:

In [21]:
type(data['size_bytes'].values)

numpy.ndarray

Больше про pandas можно найти по этом полезным ссылкам:

* Официальные туториалы: http://pandas.pydata.org/pandas-docs/stable/tutorials.html

* Статья на Хабре от OpenDataScience сообщества: https://habr.com/company/ods/blog/322626/

* Подробный гайд: https://media.readthedocs.org/pdf/pandasguide/latest/pandasguide.pdf

Главное в работе с новыми библиотеками -- не бояться тыкать в разные функции, смотреть типы возвращаемых объектов и активно пользоваться Яндексом, а ещё лучше понимать всё из docstring'а (`Shift+Tab` при нахождении курсора внутри скобок функции).